### PR TITLE
rgw: return valid Location element, CompleteMultipartUpload

### PR DESCRIPTION
**Pull request by mattbenjamin**
Returns URIs that should reach the server regardless of domain
configuration, and following tenant:bucket convention, if
applicable.

Fixes: https://tracker.ceph.com/issues/22655

Signed-off-by: Matt Benjamin <mbenjamin@redhat.com>

### CODE REVIEW

File: src/rgw/rgw_rest_s3.cc [hunk 1]
```diff
@@ -2541,20 +2541,30 @@ void RGWCompleteMultipart_ObjStore_S3::send_response()
...
+    std::string domain = (!s->info.domain.empty()) ? s->info.domain :
+      [&]() -> std::string {
+      std::string domain = "http://";
+      if (this->s->info.env->exists("SERVER_NAME")) {
+	domain.append(this->s->info.env->get("SERVER_NAME", "<SERVER_NAME>"));
+      } else {
+	domain.append(this->s->info.env->get("HTTP_HOST", "<HTTP_HOST>"));
       }
+      return domain;
+    }();
```
Comment (rzarzynski) [inline src/rgw/rgw_rest_s3.cc]: What's the reason for wrapping the code with the immediately executed lambda?

Comment (mattbenjamin) [reply src/rgw/rgw_rest_s3.cc]: just permits lining it into the ternary assignment expr.  most likely this should be consolidated into a routine, but I don't know where all it should be used yet

File: src/rgw/rgw_rest_s3.cc [hunk 2]
```diff
@@ -2541,20 +2541,30 @@ void RGWCompleteMultipart_ObjStore_S3::send_response()
...
     s->formatter->open_object_section_in_ns("CompleteMultipartUploadResult", XMLNS_AWS_S3);
-    if (!s->bucket_tenant.empty()) {
-      if (s->info.domain.length()) {
-        s->formatter->dump_format("Location", "%s.%s.%s",
-          s->bucket_name.c_str(),
-          s->bucket_tenant.c_str(),
-          s->info.domain.c_str());
+    std::string domain = (!s->info.domain.empty()) ? s->info.domain :
+      [&]() -> std::string {
+      std::string domain = "http://";
```
Comment (rzarzynski) [inline src/rgw/rgw_rest_s3.cc]: Shouldn't we consider HTTPS here?

File: src/rgw/rgw_rest_s3.cc (hunk 1 above)
Comment (rzarzynski) [reply src/rgw/rgw_rest_s3.cc]: Yeah, having a routine for that sounds nice. We might also want to unify with `dump_uri_from_state` of `rgw_rest.cc`.

File: src/rgw/rgw_rest_s3.cc (hunk 2 above)
Comment (mattbenjamin) [reply src/rgw/rgw_rest_s3.cc]: Yes.  I thought of that after typing.  And we make this same mistake elsewhere.  But are we supposed to just pick http/s, or match the incoming proto?

File: src/rgw/rgw_rest_s3.cc (hunk 1 above)
Comment (mattbenjamin) [reply src/rgw/rgw_rest_s3.cc]: does my routine (based on the code formerly in this location but adjusted for tenant) do the right thing?  dump_uri_from_state() is different, ignores tenants, deals with domain differently;  need to get it right for all cases...

File: src/rgw/rgw_rest_s3.cc (hunk 2 above)
Comment (rzarzynski) [reply src/rgw/rgw_rest_s3.cc]: We should stick to the incoming proto, I bet.

File: src/rgw/rgw_rest_s3.cc (hunk 1 above)
Comment (rzarzynski) [reply src/rgw/rgw_rest_s3.cc]: `<Location>` crafted in the new way is definitely more consistent with examples from the AWS doc. The last thing I see is the HTTPS case.

Actually `dump_uri_from_state` looks like a dead code. I will make a patch to drop it.

Comment (mattbenjamin) [reply src/rgw/rgw_rest_s3.cc]: @rzarzynski that's a big help, I'll update this one with the cleanups

**Comment by mattbenjamin** [comment]
http://pulpito.ceph.com/mbenjamin-2018-02-01_21:22:30-rgw-wip-rgw-mploc---basic-smithi/
(passed w/2 valgrind failures)

**Pull request merged**